signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+        self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._suspend_sync():
+            self._data[key] = self.from_base(value, self)
+        self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def __reversed__(self):
+        self.load()
+        return reversed(self._data)
+
+    def __iadd__(self, iterable):
+        self.load()
+        self._data += [self.from_base(value, self) for value in iterable]
+        self.sync()
+
+    def insert(self, index, item):
+        self.load()
+        with self._suspend_sync():
+            self._data.insert(index, self.from_base(item, self))
+        self.sync()
+
+    def append(self, item):
+        self.load()
+        with self._suspend_sync():
+            self._data.append(self.from_base(item, self))
+        self.sync()
+
+    def extend(self, iterable):
+        self.load()
+        with self._suspend_sync():
+            self._data.extend([self.from_base(value, self) for value in iterable])
+        self.sync()
+
+    def remove(self, item):
+        self.load()
+        with self._suspend_sync():
+            self._data.remove(self.from_base(item, self))
+        self.sync()
+
+    def clear(self):
+        self._data = []
+        self.sync()
+
+
+class JSONCollection(SyncedCollection):
+
+    def __init__(self, filename, write_concern=False):
+        self._filename = os.path.realpath(filename)
+        self._write_concern = write_concern
+
+    def _load_from_disk(self):
+        try:
+            with open(self._filename, 'rb') as file:
+                blob = file.read()
+                return json.loads(blob.decode())
+        except IOError as error:
+            if error.errno == errno.ENOENT:
+                return None
+
+    def _load(self):
+        data = self._load_from_disk()
+        with self._suspend_sync():
+            self.reset(data)
+
+    def _sync(self):
+        data = self.to_base()
+
+        # Serialize data:
+        print(data)
+        blob = json.dumps(data).encode()
+
+        if self._write_concern:
+            dirname, filename = os.path.split(self._filename)
+            fn_tmp = os.path.join(dirname, '._{uid}_{fn}'.format(
+                uid=uuid.uuid4(), fn=filename))
+            with open(fn_tmp, 'wb') as tmpfile:
+                tmpfile.write(blob)
+            os.replace(fn_tmp, self._filename)
+        else:
+            with open(self._filename, 'wb') as file:
+                file.write(blob)
+
+
+class JSONDict(JSONCollection, SyncedDict):
+    def __init__(self, filename, data=None, write_concern=False, parent=None):
+        super(JSONDict, self).__init__(filename, write_concern=write_concern)
+        super(JSONCollection, self).__init__(data=data, parent=parent)
```

Comment:
To avoid having to hack through the MRO this way, I would recommend making all the parent class `__init__` methods accept (and ignore) arbitrary kwargs so that you can just call `super().__init__(...)` and have the arguments interpreted appropriately

Reply:
@vishav1771 you'll also want to do this on the base `SyncedCollection` class, it looks like you've done it everywhere else though 👍

Reply:
I left that intentionally so that it would raise a warning when not required `kwargs` is passed.

Reply:
@vishav1771 @vyasr I looked at the code and I'm unsure if this conversation can be resolved. Currently the `SyncedCollection` class has the constructor signature: `def __init__(self, parent=None):`. Do we need to accept `**kwargs` there? Apologies in advance if I've missed another conversation on this topic. https://github.com/glotzerlab/signac/pull/336/files#diff-89d6453841b1887a33032d47f0728d7cR28

Reply:
I was talking about the warning when we pass `JSONDict('test.json', foo=1)`. IMO, We don't need to accept `kwargs` for `SyncedCollection`. As we don't pass  `kwargs` to any upper level (class). So, we have a definite set of `kwargs` that we allow for `SyncedCollection` which is only `parent` in this case.